### DISCUSSION TITLE
HOWTO: Delete a subword on Backspace

### DISCUSSION THREAD

**Discussion by ineu** [nushell/nushell#8859] [Q&A]
This belongs more to reedline, but that repo has no discussions, so posting the question here.

In zsh I used to remove subwords (i.e. parts of a word separated by underscores) using `C-w` or `M-Backspace`. I would like to replicate the same behavior in Nu. By default Nu removes the whole word which is inconvenient when you have a_very_long_identifier_to_edit.

I'm pretty sure this is configurable (probably in config.keybindings) but not really sure where to find a list of keys and commands, and what is the proper command for my case.

**Comment by ghost** [comment]
There dosnt seem to be a default.

However: you can define your own keybindings  
- [Documentation](https://www.nushell.sh/book/line_editor.html)
- [Examples](https://github.com/nushell/nushell/blob/8ee52b6ee1cebfbb265d68efcffed97254a7c252/crates/nu-utils/src/sample_config/default_config.nu#L323)
- [List of available keybindings](https://github.com/nushell/nushell/blob/8ee52b6ee1cebfbb265d68efcffed97254a7c252/crates/nu-cli/src/reedline_config.rs#L872)

---

a solution for adding it (untested):

`env.nu`:
```nu
let-env config = {
  # ... your other config stuff ...

  keybindings: [
    {
      name: delete_word
      modifier: control
      keycode: backspace
      mode: emacs
      event: { send: backspaceword }
    }
  ]
}
```

**Reply by ineu** [reply]
Thanks for the directions.
Looks like there's a bug though. I tried using `cutwordleft` and `cutbigwordleft` and both behave the same: they remove the whole token, including underscores. So it seems there's no difference between the two. That's the snippet I used for testing:
```
    {
      name:     delete_word
      modifier: control
      keycode:  char_g
      mode:     emacs
      event:    { edit: cutwordleft }
    }
```
`backspaceword` behaves the same btw.

**Reply by ghost** [reply]
ive tested around a bit - it seems the `keybinds` configuration does exactly nothing. i cant change any key to anything. maybe a bug or im just misunderstanding the docs.

i tried binding: `shift+char_m` `control+char_m` `control+backspace` `control+char_q`  
to: `{edit: moveleft}` `{edit: cutwordleft}` `{edit: movetolinestart}`  
with the mode setting `[emacs vi_insert]` since i use vi mode
and not one of them changed anything (control+m with and without binding for some reason send a newline btw)

**Reply by fdncred** [reply]
The first step to understanding if keybindings work is to understand if your terminal is interpreting them correctly. For this, nushell provides `keybindings listen`. If your terminal doesn't "hear" a keybinding, neither will nushell.

It's always possible that we have bugs, but it would surprise me if "keybinds configuration does exactly nothing", since I have and use keybindings every day with success.

**Reply by ghost** [reply]
it seems to interpret the keybinding (but wrong) (vim handles the same keybindings correct):

my input | nu intepretation | note
----- | ---- | ---
`crtl+m` |  `code: Enter, modifier: NONE, flags: 0b000000`
`m` | `char: m, code: 0x00006d, modifier: NONE, flags: 0b000000`
`crtl+o` | `char: o, code: 0x00006f, modifier: CONTROL, flags: 0b000010` | i moved this key via xmodmap
`crtl+backspace` | `char: h, code: 0x000068, modifier: CONTROL, flags: 0b000010`
`crtl+return` | `code: Enter, modifier: NONE, flags: 0b000000`
`crtl+i` | `code: Tab, modifier: NONE, flags: 0b000000`

but a few also work..

ill look into this seperately and maybe open a issue if it is persistent and not my fault

**Comment by ineu** [comment]
I looked at the code and it seems the problem here is not a keybinding. Actually alt+backspace is already bound to the proper action:
```
~ ❯ keybindings default | find word | get 0
╭──────────┬───────────────────────╮
│ mode     │ emacs                 │
│ modifier │ KeyModifiers(ALT)     │
│ code     │ Backspace             │
│ event    │ Edit([BackspaceWord]) │
╰──────────┴───────────────────────╯
```

The real problem is the notion of "word" in Nu. I looked at the code and it seems that "word" means a unicode word. So alt+backspace deletes to the left to the closest whitespace or dash, which would make sense for the natural languages.

But a shell is no natural language. Thus I believe the separator list must also include underscore and dot characters.

Alas, I couldn't add it in the keybindings:
- `event:    { edit: cutleftuntil, value: "." }` only accepts one character in a value
- `until` only works with menus, as described in https://www.nushell.sh/book/line_editor.html#until-type

**Reply by fdncred** [reply]
If this doesn't work like it should, reedline may need to be changed.

**Reply by ineu** [reply]
It depends on the purpose of reedline. If it only exists to be used in the shell, then yes, it should be changed. On the other hand, if it's a general-purpose readline replacement, then it's up to a consumer to define the meaning of "word". For example, readline is used not only for bash, but also for psql. Shell script and SQL are very different languages with different semantics.
So I don't consider this a bug in reedline. But maybe a feature request to make "word" configurable.

**Reply by fdncred** [reply]
There are so many fiddly bits when it comes to vim keybindings/vim motions. I wish there was an official vim-motions crate that people could rely on. Having said that, it wouldn't fix the need for redefinition of things. That sounds like one could have a specific config file that defines what you mean by "word" and other vim terms.
